docs/designs/high_performance_ps.md

```diff
@@ -0,0 +1,63 @@
+# High Performance Parameter Server Design
+
+## Motivation
+
+This design doc focus on implementing a high performance parameter server(short for PS). For the functionality of the PS, please refer to this [design doc](https://github.com/sql-machine-learning/elasticdl/blob/develop/docs/designs/parameter_server.md)
+
+PS receives gradients from workers, applies gradients to parameters, and sends the latest parameters to workers. Receiving gradients and sending parameters bring IO workload to PS, and applying gradients to parameters brings CPU workload to PS. Since one PS could receive gradients from many workers, both IO workload and CPU workload would be very heavy.
+
+The current PS is implemented with Python. Because of [GIL](https://wiki.python.org/moin/GlobalInterpreterLock) of Python, gradients are applied to parameters sequentially with only one CPU core. As a result, the receiving gradients service is also blocked, and waiting for current gradients to be consumed. To resolve this bottleneck, we have to fully use multi CPU cores of PS.
```

Comment:
As a result, the receiving gradients service is also blocked -> The receiving gradients service is not blocked,  but will only service when the running thread for applying gradients is preempted.  It cannot service in parallel.

Reply:
Currently, the receiving gradients and applying gradients are in the same thread. It seems that if applying gradients is not finished, the thread could not provide receiving gradients service.